test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java

```diff
@@ -104,6 +106,7 @@ public class MockTransportService extends TransportService {
     private final Map<DiscoveryNode, List<Transport.Connection>> openConnections = new HashMap<>();
 
     private final List<Runnable> onStopListeners = new CopyOnWriteArrayList<>();
+    private final AtomicReference<Consumer<Transport.Connection>> onConnectionClosedCallback = new AtomicReference<>();
```

Comment:
It does not look like we need CAS operations on this field.
May be this could be simplified to volatile field or even a regular one if not expected to be used from different threads?

Reply:
We need visibility guarantee. So `volatile` can work. You are right that we don't really need CAS though I have a minor preference to have this field as final to align with others.